src/openrct2/scripting/ScSocketServer.hpp

```diff
@@ -0,0 +1,505 @@
+        {
+            return _plugin;
+        }
+
+        virtual void Update() = 0;
+
+        virtual void Dispose()
+        {
+        }
+
+        virtual bool IsDisposed() const = 0;
+    };
+
+    class ScSocket : public ScSocketBase
+    {
+    private:
+        static constexpr uint32_t EVENT_NONE = std::numeric_limits<uint32_t>::max();
+        static constexpr uint32_t EVENT_CLOSE = 0;
+        static constexpr uint32_t EVENT_DATA = 1;
+        static constexpr uint32_t EVENT_CONNECT_ONCE = 2;
+        static constexpr uint32_t EVENT_ERROR = 3;
+
+        EventList _eventList;
+        std::unique_ptr<ITcpSocket> _socket;
+        bool _disposed{};
+        bool _connecting{};
+
+    public:
+        ScSocket(const std::shared_ptr<Plugin>& plugin)
+            : ScSocketBase(plugin)
+        {
+        }
+
+        ScSocket(const std::shared_ptr<Plugin>& plugin, std::unique_ptr<ITcpSocket>&& socket)
+            : ScSocketBase(plugin)
+            , _socket(std::move(socket))
+        {
+        }
+
+    private:
+        ScSocket* destroy(const DukValue& error)
+        {
+            if (_socket != nullptr)
+            {
+                _socket->Close();
+                _socket = nullptr;
+            }
+            return this;
+        }
+
+        ScSocket* setNoDelay(bool noDelay)
+        {
+            if (_socket != nullptr)
+            {
+                _socket->SetNoDelay(noDelay);
+            }
+            return this;
+        }
+
+        ScSocket* connect(uint16_t port, const std::string& host, const DukValue& callback)
+        {
+            auto ctx = GetContext()->GetScriptEngine().GetContext();
+            if (_socket != nullptr)
+            {
+                duk_error(ctx, DUK_ERR_ERROR, "Socket has already been created.");
+            }
+            else if (_disposed)
+            {
+                duk_error(ctx, DUK_ERR_ERROR, "Socket is disposed.");
+            }
+            else if (_connecting)
+            {
+                duk_error(ctx, DUK_ERR_ERROR, "Socket is already connecting.");
+            }
+            else if (!IsLocalhostAddress(host))
+            {
+                duk_error(ctx, DUK_ERR_ERROR, "For security reasons, only connecting to localhost is allowed.");
+            }
+            else
+            {
+                _socket = CreateTcpSocket();
+                _socket->Connect(host, port);
+                _eventList.AddListener(EVENT_CONNECT_ONCE, callback);
+                _connecting = true;
+            }
+            return this;
+        }
+
+        ScSocket* end(const DukValue& data)
+        {
+            if (_disposed)
+            {
+                auto ctx = GetContext()->GetScriptEngine().GetContext();
+                duk_error(ctx, DUK_ERR_ERROR, "Socket is disposed.");
+            }
+            else if (_socket != nullptr)
+            {
+                if (data.type() == DukValue::Type::STRING)
```

Comment:
Passing any other type will have no-op, this should be made clear by an error or support for those types.